components/script/dom/htmlimageelement.rs

```diff
@@ -789,26 +791,38 @@ impl HTMLImageElement {
         // Step 2 abort if user-agent does not supports images
         // NOTE: Servo only supports images, skipping this step
 
-        // step 3, 4
-        // TODO: take srcset and parent images into account
-        if !src.is_empty() {
-            // TODO: take pixel density into account
+        // Step 3, 4
+        let mut selected_source = None;
+        let mut pixel_density = None;
+        let src_set = elem.get_string_attribute(&local_name!("srcset"));
+        let is_parent_picture = elem.upcast::<Node>().GetParentElement().unwrap().is::<HTMLPictureElement>();
+        if src_set.is_empty() && !is_parent_picture &&  !src.is_empty() {
+            selected_source = Some(src.clone());
+            pixel_density = Some(1 as f64);
+        };
+
+        //Step 5
+        *self.last_selected_source.borrow_mut() = selected_source.clone();
+
+        // Step 6, check the list of available images
+        if !selected_source.unwrap().is_empty() {
```

Comment:
Unwrapping here will panic for any image that has a srcset or a picture parent. We should use `.as_ref().map_or(false, |source| source.is_empty())`.